lib/global-styles.php

```diff
@@ -34,98 +44,299 @@ function gutenberg_get_css_vars( $global_styles_branch, $prefix = '', $is_start
+ */
+function gutenberg_experimental_global_styles_get_user_cpt_id() {
+	$user_cpt_id = null;
+	$user_cpt    = gutenberg_experimental_global_styles_get_user_cpt( array( 'publish', 'draft' ), true );
+	if ( array_key_exists( 'ID', $user_cpt ) ) {
+		$user_cpt_id = $user_cpt['ID'];
 	}
+	return $user_cpt_id;
+}
+
+/**
+ * Return core's Global Styles.
+ *
+ * @return array Global Styles tree.
+ */
+function gutenberg_experimental_global_styles_get_core() {
+	return gutenberg_experimental_global_styles_get_from_file(
+		dirname( dirname( __FILE__ ) ) . '/experimental-default-global-styles.json'
+	);
+}
+
+/**
+ * Return theme's Global Styles.
+ *
+ * @return array Global Styles tree.
+ */
+function gutenberg_experimental_global_styles_get_theme() {
+	return gutenberg_experimental_global_styles_get_from_file(
+		locate_template( 'experimental-theme.json' )
+	);
+}
+
+/**
+ * Takes a Global Styles tree and returns a CSS rule
+ * that contains the corresponding CSS custom properties.
+ *
+ * @param array $global_styles Global Styles tree.
+ * @return string CSS rule.
+ */
+function gutenberg_experimental_global_styles_resolver( $global_styles ) {
+	$css_rule = '';
 
+	$token    = '--';
+	$prefix   = '--wp' . $token;
+	$css_vars = gutenberg_experimental_global_styles_get_css_vars( $global_styles, $prefix, $token );
 	if ( empty( $css_vars ) ) {
-		return;
+		return $css_rule;
 	}
 
-	$inline_style = ":root {\n";
+	$css_rule = ":root {\n";
 	foreach ( $css_vars as $var => $value ) {
-		$inline_style .= "\t" . $var . ': ' . $value . ";\n";
+		$css_rule .= "\t" . $var . ': ' . $value . ";\n";
+	}
+	$css_rule .= '}';
+
+	return $css_rule;
+}
+
+/**
+ * Fetches the Global Styles for each level (core, theme, user)
+ * and enqueues the resulting CSS custom properties for the editor.
+ */
+function gutenberg_experimental_global_styles_enqueue_assets_editor() {
+	if ( gutenberg_experimental_global_styles_is_site_editor() ) {
+		gutenberg_experimental_global_styles_enqueue_assets();
+	}
+}
+
+/**
+ * Fetches the Global Styles for each level (core, theme, user)
+ * and enqueues the resulting CSS custom properties.
+ */
+function gutenberg_experimental_global_styles_enqueue_assets() {
+	if ( ! gutenberg_experimental_global_styles_has_theme_support() ) {
+		return;
+	}
+
+	$global_styles = array_merge(
+		gutenberg_experimental_global_styles_get_core(),
+		gutenberg_experimental_global_styles_get_theme(),
+		gutenberg_experimental_global_styles_get_user()
+	);
+
+	$inline_style = gutenberg_experimental_global_styles_resolver( $global_styles );
+	if ( empty( $inline_style ) ) {
+		return;
 	}
-	$inline_style .= '}';
 
 	wp_register_style( 'global-styles', false, array(), true, true );
 	wp_add_inline_style( 'global-styles', $inline_style );
 	wp_enqueue_style( 'global-styles' );
 }
-add_action( 'enqueue_block_assets', 'gutenberg_enqueue_global_styles_assets' );
 
 /**
- * Adds class wp-gs to the frontend body class if the theme defines a experimental-theme.json.
```

Comment:
If we unconditionally add the wp-gs class we will break the styles of some blocks.
```
.wp-gs .wp-block-button__link:not(.has-background) {
	background-color: var(--wp--color--primary);
}
```

As the class is added even if the theme does not support global styles a theme that was changing the button color using the following code will stop working:
```
 .wp-block-button__link {
	background-color: red;
}
```

I think the whole idea of adding wp-gs was to mark themes that support global styles if we unconditionally add the class we would be able to not add it at all as well.
I know that for now we only add if the experiment is enabled but one day we remove the experiment and in that case, we would unconditionally add the class.
I think we should add the class when the experiment is enabled and when the theme supports global styles.

Reply:
We already do that! The class is only added when the experiment is enabled, the theme has support, and the editor is site editor (for back-end, doesn't apply to front-end).